mesh/meshdb.go

```diff
@@ -127,36 +126,49 @@ func (m *MeshDB) LayerBlockIds(index types.LayerID) ([]types.BlockID, error) {
 	return blockids, nil
 }
 
-func (mc *MeshDB) ForBlockInView(view map[types.BlockID]struct{}, layer types.LayerID, blockHandler func(block *types.BlockHeader) error) error {
-	stack := list.New()
-	for b := range view {
-		stack.PushFront(b)
-	}
-	set := make(map[types.BlockID]struct{})
-	for b := stack.Front(); b != nil; b = stack.Front() {
-		a := stack.Remove(stack.Front()).(types.BlockID)
+type blockIdQueue struct {
+	l []types.BlockID
+}
+
+func (s *blockIdQueue) Push(b types.BlockID) {
+	s.l = append(s.l, b)
+}
+
+func (s *blockIdQueue) Pop() types.BlockID {
+	b := s.l[0]
+	s.l = s.l[1:]
+	return b
+}
+
+func (s *blockIdQueue) IsEmpty() bool {
+	return len(s.l) == 0
+}
 
-		block, err := mc.GetBlock(a)
+func (m *MeshDB) ForBlockInView(view map[types.BlockID]struct{}, layer types.LayerID, blockHandler func(block *types.Block) error) error {
+	blocksToVisit := &blockIdQueue{}
+	for id := range view {
+		blocksToVisit.Push(id)
+	}
+	seenBlocks := make(map[types.BlockID]struct{})
+	for !blocksToVisit.IsEmpty() {
+		block, err := m.GetBlock(blocksToVisit.Pop())
 		if err != nil {
 			return err
 		}
+		if block.LayerIndex < layer {
+			continue // dont traverse too deep
+		}
```

Comment:
dont put blocks that are to deep in the queue in the first place

Reply:
This is an intentional change and tradeoff. The alternative is to get the block from the cache twice (once to check the layer and another to perform processing, incl. getting the children).

The cost of checking this at this point is that we need to enter the loop more times. Specifically, we're expected to enter the loop about 200 more times (an additional layer) but save `200 * <layers traversed>` calls to `getBlock()`. Empirically speaking, it results in reduced runtime in the benchmark (10-15%).

Reply:
the blocks in the second "GetBlock" are the blocks from the view so what you say about calling it twice on the same block  is incorrect 
i think validating the blocks once they come out of the queue is better

Reply:
"the blocks from the view" are what gets put into the queue and then we call `GetBlock()` on them when it's their turn to be taken out of the queue.

I agree that it would be cleaner, but I don't think it's worth a 10-15% performance hit.

Reply:
a better way to fix it is to out pointers to blocks in the queue instead of id's

Reply:
in your solution the queue will hold blocks that are not needed at all 
maybe a better way to fix it is to put pointers to blocks in the queue instead of id's
keep the same -> flow avoid the double get

Reply:
This was my initial thought as well and if you look at the commits in this PR you'll find this commit that does that: https://github.com/spacemeshos/go-spacemesh/pull/1212/commits/cdb68faa74f4f6593f67f8194ca0f81234cfe001

I later reverted this change and changed the flow because I came to the realization that the memory footprint of keeping all these blocks would be huge:

200 blocks per layer
2 layers in the queue (after handling the first block you should have all of the next layer queued, while still having most of the first layer there)
Let's take an optimistic approach and say a mini block is 10kb.

400 blocks * 10kb = 4mb

This is a very conservative calculation and it's still too big to run this in multiple threads concurrently. WDYT?

Reply:
these blocks are already in the queue you are just holding a pointer i dont agree with the reasons or the calculations you gave

Reply:
besides you know that the all the block ids in the view are from layers older then the block you are working on you don't need to pull them from the cache to figure out if you should insert the id in the queue.
correct me if im wrong but in this way you get only one call to getBlock
and the condition can stay before the insertion.

Reply:
Great point! I've added this check now in 6f857771. However, we still need the check before the handler because blocks could reference blocks more than one layer back and then blocks could "slip through".

If you still don't like the way it's implemented then let's discuss this in person tomorrow.

Reply:
nope you did just what i mean it's great

Reply:
Perfect! Yuval asked me to get an approval from you before merging -- any other reservations?

Reply:
thanks, already approved it